text/0000-packages-as-optional-namespaces.md

```diff
@@ -0,0 +1,230 @@
+
+# Motivation
+
+While Rust crates are practically unlimited in size, it is a common pattern for organizations to split their projects into many crates, especially if they expect users to only need a fraction of their crates or they have different backwards compatibility guarantees.
+
+For example, [unic](https://crates.io/search?page=1&per_page=10&q=unic-), [tokio](https://crates.io/search?page=1&per_page=10&q=tokio-), [async-std](https://crates.io/search?page=1&per_page=10&q=async-), [rusoto](https://crates.io/search?q=rusoto) all do something like this, with lots of `projectname-foo` crates. At the moment, it is not necessarily true that a crate named `projectname-foo` is maintained by `projectname`, and in some cases that is even desired! E.g. `serde` has many third party "plugin" crates like [serde-xml-rs](https://github.com/RReverser/serde-xml-rs). Similarly, [async-tls](https://crates.io/crates/async-tls) is a general crate not specific to the async-std ecosystem.
+
+Regardless, it is nice to have a way to signify "these are all crates belonging to a single project, and you may trust them the same" and discover these related crates. When starting up [ICU4X](https://github.com/unicode-org/icu4x/), we came up against this problem: We wanted to be able to publish ICU4X as an extremely modular system of `icu-foo` or `icu4x-foo` crates, but it would be confusing to users if third-party crates could also exist there (or take names we wanted to use).
+
+It's worth spending a bit of time talking about "projects" and "organizations", as nebulous as those terms are. This feature is *primarily* motivated by the needs of "projects". By this I mean a _single_ Rust API developed as multiple crates, for example `serde` and `serde::derive`, or `icu` and `icu::provider`, or `servo::script` and `servo::layout`. One would expect "projects" like this to live under a single Git repository according to the norms of project organization; they are logically a single project and API even if they are multiple crates.
+
+The feature suggested here is unlikely to be used by "organizations" as this would put independent concerns in the same Rust API.  By "organizations", I mean a group of people who are coming together to build likely related crates, under the same "brand", likely developed in multiple repos under a GitHub organization.
+
+
+The motivation here is distinct from the general problem of squatting -- with general squatting, someone else might come up with a cool crate name before you do. However, with `projectname-foo` crates, it's more of a case of third parties "muscling in" on a name you have already chosen and are using.
+
+# Guide-level explanation
+
+The owners of the `foo` crate may provide other crates under the `foo` namespace, like `foo::bar`.  For users, this makes its official status clearer and makes it easier to discover.
+
+Users import these crates in Cargo.toml as normal:
+
+```toml
+[dependencies]
+"foo" = "1.0.42"
+"foo::bar" = "3.1"
```

Comment:
Are there any special considerations for rustdoc we should keep in mind with this feature?

Reply:
I think rustdoc may need some code changes for this to work; but overall I think that it ought to be fine? it might choose to render these slightly differently.

Reply:
(sorry for the necro post)

Locally, we are treating them as one package but they won't find that by browsing rustdoc, locally or via docs.rs with a naiive solution.  My concern is that this would be like `features` (without `autocfg`) where it can be weird correlating everything which would impact understandability of a code base / example code.

Random ideas:
- When building locally, show `foo::bar` as a ``mod bar` in `foo` with a visual tag that this is the `foo::bar` dependency
- More docs.rs, should have a way to link these together for discovery of `foo::bar` from `foo` or to browse back to `foo`
  - As a cheap idea, what if we could do `#[cfg_attr(docsrs, rustdoc(namespaced))] mod bar {}` and then `rustdoc` would make that `mod` a link.

Reply:
We'll need to figure out how namespaced names are written to the filesystem for rustdoc, and the mechanism that rustdoc will use to find them.

Currently, rustdoc hunts for a directory with the same name as the library. However, we can't use `foo::bar` as a directory name (Windows), and we can't use just `bar` (it is not sufficiently unique). Other normalizations might have problems (like `foo_bar` could also collide with a real `foo_bar`). The directories could be nested, but I don't know how complicated that will be for [`ExternalCrate::location`](https://github.com/rust-lang/rust/blob/f81d6f0cb00b1e314107b81bcb9aa72b9df9f564/src/librustdoc/clean/types.rs#L161-L169).

Cargo could also start using something like `--extern-html-root-url`, but unfortunately that is hung up on doing more design work (see https://github.com/rust-lang/cargo/issues/10241#issuecomment-1003761078 and https://github.com/rust-lang/cargo/issues/8296).